### PR TITLE
[core] Avoid shared pointer capture health check

### PR DESCRIPTION
**Pull request by dentiny**
Address https://github.com/ray-project/ray/pull/49471/files/c4c3b4aa16ce8bb263582e1edd4d40ba6fc722d2#r1899285992
Already checked with TSAN, no complaints for data race.

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc [hunk 1]
```diff
@@ -158,13 +158,20 @@ void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
...
+        }
+
         // This callback is done in gRPC's thread pool.
         STATS_health_check_rpc_latency_ms.Record(
             absl::ToInt64Milliseconds(absl::Now() - start));
+
+        // Have to capture shared pointer for health check manager to ensure `io_service`
+        // access is valid.
         manager->io_service_.post(
             [this, status, manager]() {
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_health_check_manager.cc:176]: Same here.

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc [hunk 2]
```diff
@@ -158,13 +158,20 @@ void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
...
   // pointer to make sure all memory accesses are valid.
   stub_->async()->Check(
-      &context_,
-      &request_,
-      &response_,
-      [this, start = now, manager = manager](::grpc::Status status) {
+      &context_, &request_, &response_, [this, start = now](::grpc::Status status) {
+        auto manager = manager_.lock();
+        if (manager == nullptr) {
+          RAY_CHECK(stopped_);
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_health_check_manager.cc:164]: Actually is this always true? I don't see the destructor removes all nodes.

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc (hunk 1 above)
Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:176]: No, you cannot make sure `io_context` access is valid if you don't capture `manager`.

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:176]: Checkout my previous reply here: https://github.com/ray-project/ray/pull/49471#discussion_r1899058303

File: src/ray/gcs/gcs_server/gcs_health_check_manager.cc (hunk 2 above)
Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:164]: https://github.com/ray-project/ray/blob/66cec2f36fcf1582a53fcb54564936150e602f09/src/ray/gcs/gcs_server/gcs_health_check_manager.cc#L208

It's updated to true when health check manager requests to remove node.

Comment (jjyao) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc:164]: Yea but `RemoveNode` is not guaranteed to be called when the manager is destructed?

Comment (dentiny) [reply src/ray/gcs/gcs_server/gcs_health_check_manager.cc]: Gotcha, removed the check.

**Review by jjyao — APPROVED** [review]

**Comment by jjyao** [comment]
Merge conflicts

**Pull request merged**